prow/plugins/wip/wip-label.go

```diff
@@ -0,0 +1,184 @@
+*/
+
+// Package wip will label a PR a work-in-progress if the author provides
+// a prefix to their pull request title to the same effect. The submit-
+// queue will not merge pull requests with the work-in-progress label.
+// The label will be removed when the title changes to no longer begin
+// with the prefix.
+package wip
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	wipLabel   = "do-not-merge/work-in-progress"
+	pluginName = "wip"
+)
+
+var (
+	wipTitlePrefix = regexp.MustCompile(`(?i)^[\[<]?WIP[\]>]?`)
+	addingWipLabel = fmt.Sprintf("Labelling this PR as %s due to the title.", wipLabel)
+)
+
+// detailsForDevs generates an explanation of what this bot
+// is doing for developers so that they can understand how
+// to interact with it
+func detailsForDevs(title string) string {
+	format := `Your pull request title starts with %q, so the %s label will be added.
+
+This label will ensure that your pull request will not be merged. Remove the prefix from your pull request title to trigger the removal of the label and allow for your pull request to be merged.
+`
+	prefix := wipTitlePrefix.FindString(title)
+	return fmt.Sprintf(format, prefix, wipLabel)
+}
+
+type event struct {
+	org                string
+	repo               string
+	number             int
+	hasWipLabel        bool
+	needsWipLabel      bool
+	hasWipLabelComment bool
+	wipLabelCommentID  int
+	wipLabelBody       string
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	CreateComment(org, repo string, number int, comment string) error
+	DeleteComment(org, repo string, ID int) error
+}
+
+func handlePullRequest(pc plugins.PluginClient, pe github.PullRequestEvent) error {
+	// These are the only actions indicating the PR title may have changed.
+	if pe.Action != github.PullRequestActionOpened && pe.Action != github.PullRequestActionEdited && pe.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+
+	var (
+		org    = pe.PullRequest.Base.Repo.Owner.Login
+		repo   = pe.PullRequest.Base.Repo.Name
+		number = pe.PullRequest.Number
+		author = pe.PullRequest.User.Login
+		title  = pe.PullRequest.Title
+	)
+
+	currentLabels, err := pc.GitHubClient.GetIssueLabels(org, repo, number)
+	if err != nil {
+		return fmt.Errorf("could not get labels for PR %s/%s:%d in WIP plugin: %v", org, repo, number, err)
+	}
+	hasWipLabel := hasWipLabel(currentLabels)
+
+	needsWipLabel := hasWipPrefix(title)
+
+	wipLabelBody := plugins.FormatResponse(author, addingWipLabel, detailsForDevs(title))
+	currentComments, err := pc.GitHubClient.ListIssueComments(org, repo, number)
```

Comment:
How many github tokens will this use? I suspect it will increase prow's total token usage by an order of magnitude.

Reply:
Hard to say. Most PRs have fewer than 100 comments, so I'd expect this to be impactful but ultimately not an order of magnitude. I don't see a better way to deal with this unless we:
 - make comment addition best-effort
 - don't care to clean up comments

Reply:
Or don't write a comment at all

Reply:
Ah, I was thinking about it from the point of view that you'd be using a token to add/remove the label and list labels anyway, so this would be +1 token at best. However, the argument is valid. We can just always add the comment even if one already exists and never remove it, that doesn't strike me as bad.